cmd/dra-example-kubeletplugin/health.go

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright 2025 The Kubernetes Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/url"
+	"strconv"
+	"sync"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/grpc/health/grpc_health_v1"
+	"google.golang.org/grpc/status"
+	"k8s.io/klog/v2"
+	drapb "k8s.io/kubelet/pkg/apis/dra/v1beta1"
+	registerapi "k8s.io/kubelet/pkg/apis/pluginregistration/v1"
+
+	"sigs.k8s.io/dra-example-driver/pkg/consts"
+)
+
+type healthcheck struct {
+	grpc_health_v1.UnimplementedHealthServer
+
+	server *grpc.Server
+	wg     sync.WaitGroup
+
+	regClient registerapi.RegistrationClient
+	draClient drapb.DRAPluginClient
+}
+
+func startHealthcheck(ctx context.Context, config *Config) (*healthcheck, error) {
+	log := klog.FromContext(ctx)
+
+	port := config.flags.healthcheckPort
+	if port == 0 {
+		return nil, nil
+	}
+
+	addr := net.JoinHostPort("", strconv.Itoa(port))
+	lis, err := net.Listen("tcp", addr)
+	if err != nil {
+		return nil, fmt.Errorf("failed to listen for healthcheck service at %s: %w", addr, err)
+	}
+
+	// TODO: this needs to adapt when seamless upgrades
+	// are enabled and the filename includes a uid.
+	regSockPath, err := url.JoinPath("unix:///", config.flags.kubeletRegistrarDirectoryPath, consts.DriverName+"-reg.sock")
+	if err != nil {
+		return nil, fmt.Errorf("invalid registration socket path: %w", err)
+	}
+	log.Info("connecting to registration socket", "path", regSockPath)
+	regConn, err := grpc.NewClient(
+		regSockPath,
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("connect to registration socket: %w", err)
+	}
+
+	draSockPath, err := url.JoinPath("unix:///", config.DriverPluginPath(), "dra.sock")
+	if err != nil {
+		return nil, fmt.Errorf("invalid DRA socket path: %w", err)
+	}
+	log.Info("connecting to DRA socket", "path", draSockPath)
+	draConn, err := grpc.NewClient(
+		draSockPath,
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("connect to DRA socket: %w", err)
+	}
+
+	server := grpc.NewServer()
+	healthcheck := &healthcheck{
+		server:    server,
+		regClient: registerapi.NewRegistrationClient(regConn),
+		draClient: drapb.NewDRAPluginClient(draConn),
+	}
+	grpc_health_v1.RegisterHealthServer(server, healthcheck)
```

Comment:
As we're registering it as a HealthServer would it make sense to implement [gRPC HealthServer](https://pkg.go.dev/google.golang.org/grpc/health/grpc_health_v1#HealthServer) interface?

Reply:
Yes, that's what this does by embedding the `grpc_health_v1.UnimplementedHealthServer` type, otherwise it wouldn't compile. The kubelet only invokes `Check` AFAICT, which is the only method we override here. `List` and `Watch` are left unimplemented and will return errors indicating that if they're called.

Reply:
I see now that we import an older version of the gRPC module which doesn't include `List` in the `HealthServer` interface: https://pkg.go.dev/google.golang.org/grpc@v1.68.1/health/grpc_health_v1#HealthServer

Since the `List` method uses defined `HealthListRequest` and `HealthListResponse` types, I don't think we can reasonably jump ahead and implement it anyway. Should I update the module in this PR to be able to do that?

Reply:
No, you shouldn't update it just for this reason.